cmd/tailscale/cli/funnel_v2.go

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package cli
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/peterbourgon/ff/v3/ffcli"
+	"tailscale.com/ipn"
+)
+
+// newFunnelCommand returns a new "funnel" subcommand using e as its environment.
+// The funnel subcommand is used to turn on/off the Funnel service.
+// Funnel is off by default.
+// Funnel allows you to publish a 'tailscale serve' server publicly, open to the
+// entire internet.
+// newFunnelCommand shares the same serveEnv as the "serve" subcommand. See
+// newServeCommand and serve.go for more details.
+func newFunnelCommand2(e *serveEnv) *ffcli.Command {
+	return &ffcli.Command{
+		Name:      "funnel",
+		ShortHelp: "Turn on/off Funnel service",
+		ShortUsage: strings.Join([]string{
+			"funnel <serve-port>",
+			"funnel <serve-port> {on|off}",
+			"funnel status [--json]",
+		}, "\n  "),
+		LongHelp: strings.Join([]string{
+			"Funnel allows you to expose your local",
+			"server publicly to the entire internet.",
+		}, "\n"),
+		Exec:      e.runFunnel2,
+		UsageFunc: usageFunc,
+		Subcommands: []*ffcli.Command{
+			{
+				Name:      "status",
+				Exec:      e.runServeStatus,
+				ShortHelp: "show current serve/funnel status",
+				FlagSet: e.newFlags("funnel-status", func(fs *flag.FlagSet) {
+					fs.BoolVar(&e.json, "json", false, "output JSON")
+				}),
+				UsageFunc: usageFunc,
+			},
+		},
+	}
+}
+
+// runFunnel is the entry point for the "tailscale funnel" subcommand and
+// manages turning on/off funnel. Funnel is off by default.
+//
+// Note: funnel is only supported on single DNS name for now. (2022-11-15)
+func (e *serveEnv) runFunnel2(ctx context.Context, args []string) error {
+	// TODO: allow length of args to be 2 when adding "on|off"
+	if len(args) != 1 {
+		return flag.ErrHelp
+	}
+	var source string
+	port64, err := strconv.ParseUint(args[0], 10, 16)
+	if err == nil {
+		source = fmt.Sprintf("http://127.0.0.1:%d", port64)
+	} else {
+		source, err = expandProxyTarget(args[0])
+	}
+	if err != nil {
+		return err
+	}
+
+	st, err := e.getLocalClientStatusWithoutPeers(ctx)
+	if err != nil {
+		return fmt.Errorf("getting client status: %w", err)
+	}
+
+	if err := e.verifyFunnelEnabled(ctx, st, 443); err != nil {
+		return err
+	}
+
+	dnsName := strings.TrimSuffix(st.Self.DNSName, ".")
+	hp := ipn.HostPort(dnsName + ":443") // TODO(marwan-at-work): support the 2 other ports
```

Comment:
Nit: maybe warn if we're overriding an existing config

Reply:
@soniaappasamy this actually brings up a question of where the config manipulation happens: on the client (cli) or on the backend (tailscaled): right now it's on the backend which has the benefit of "ensured cleanup" via the "defer" that when the connection is severed (gracefully or otherwise) we know the serve is going to be cleaned up. 

If the CLI, on the other hand, gets/sets/defers config manipulation, then it has the added benefit of being able to check/warn on pre-existing serves. But that also means a "defer" may actually not end up running if the client connection is killed ungracefully and the backend right now is not smart enough to clean only a certain part of the whole config.

So with the current implementation (config manipulation on the backend), we can either: 

1. Duplicate some of the logic and let the CLI get the serve config and check/warn. Or,
2. The backend can actually push some warnings and not just request-logs that we can show the user. 

I'm leaning towards option 2 but as a follow up. 

In the future though, it might be nice to follow the docker model: create a single serve as a separate end point, and then send another request to attach yourself to its logs. 

I'm not sure if this is something we should do soon, or if it requires some designs first.

Reply:
All of the above sounds good! ✌🏼 

> But then does add back an extra step that would be nice to eliminate in the "start up a quick proxy so I can show someone something and then tear it down" situation.

FWIW the end user would still only run a single command, so that detail/complexity would be hidden from them. Under the hood, the CLI would have to make a a couple of calls to the backend instead of one as you mentioned.